File: src/dataset/cpp/nucleotide_count_Exercism_johnngugi/gemini/GeminiFlash_nucleotide_count_v2.cpp
#include "nucleotide_count.h"
#include <stdexcept>
#include <algorithm>

using namespace dna;
using namespace std;

counter::counter(const string& a) : strand(a) {
    nucleotide_count['A'] = 0;
    nucleotide_count['C'] = 0;
    nucleotide_count['G'] = 0;
    nucleotide_count['T'] = 0;

    for (char c : a) {
        if (c != 'A' && c != 'C' && c != 'G' && c != 'T') {
            throw invalid_argument("Invalid nucleotide: " + string(1, c));
        }
        ++nucleotide_count[c];
    }
}

int counter::count(char nucleotide) const {
    if (nucleotide != 'A' && nucleotide != 'C' && nucleotide != 'G' && nucleotide != 'T') {
        throw invalid_argument("Invalid nucleotide: " + string(1, nucleotide));
    }
    return nucleotide_count.at(nucleotide);
}

map<char, int> counter::nucleotide_counts() const {
    return nucleotide_count;
}

File: src/dataset/go/nucleotide_count_Exercism_go_rootulp/gemini/GeminiFlash_nucleotide_count_v2.go
package dna

import "fmt"

// Histogram is a mapping from nucleotide to its count in given DNA.
type Histogram map[rune]int

type DNA string

// Counts generates a histogram of valid nucleotides in the given DNA.
// Returns an error if d contains an invalid nucleotide.
func (d DNA) Counts() (Histogram, error) {
	h := Histogram{
		'A': 0,
		'C': 0,
		'G': 0,
		'T': 0,
	}

	for i := 0; i < len(d); i++ {
		r := rune(d[i])
		switch r {
		case 'A', 'C', 'G', 'T':
			h[r]++
		default:
			return Histogram{}, fmt.Errorf("DNA strand %s contains invalid nucleotides", d)
		}
	}

	return h, nil
}